match/lib/match/storage/google_cloud_storage.rb

```diff
@@ -0,0 +1,245 @@
+                     google_cloud_keys_file: nil)
+        self.type = type if type
+        self.platform = platform if platform
+        self.google_cloud_keys_file = google_cloud_keys_file
+        self.bucket_name = google_cloud_bucket_name
+
+        ensure_keys_file_exists
+
+        # Extract the Project ID from the `JSON` file
+        # so the user doesn't have to provide it manually
+        keys_file_content = JSON.parse(File.read(self.google_cloud_keys_file))
+        project_id = keys_file_content["project_id"]
+        if project_id.to_s.length == 0
+          UI.user_error!("Provided keys file on path #{File.expand_path(self.google_cloud_keys_file)} doesn't include required value for `project_id`")
+        end
+
+        # Create the Google Cloud Storage client
+        # If the JSON auth file is invalid, this line will
+        # raise an exception
+        begin
+          self.gc_storage = Google::Cloud::Storage.new(
+            credentials: self.google_cloud_keys_file,
+            project_id: project_id
+          )
+        rescue => ex
+          UI.error(ex)
+          UI.verbose(ex.backtrace.join("\n"))
+          UI.user_error!("Couldn't log into your Google Cloud account using the provided JSON file at path '#{File.expand_path(self.google_cloud_keys_file)}'")
+        end
+
+        ensure_bucket_is_selected
+      end
+
+      def download
+        # Check if we already have a functional working_directory
+        return self.working_directory if @working_directory
+
+        # No existing working directory, creating a new one now
+        self.working_directory = Dir.mktmpdir
+
+        bucket.files.each do |current_file|
+          file_path = current_file.name # e.g. "N8X438SEU2/certs/distribution/XD9G7QCACF.cer"
+          download_path = File.join(self.working_directory, file_path)
+
+          FileUtils.mkdir_p(File.expand_path("..", download_path))
+          UI.verbose("Downloading file from Google Cloud Storage '#{file_path}' on bucket #{self.bucket_name}")
+          current_file.download(download_path)
+        end
+        UI.verbose("Successfully downloaded files from GCS to #{self.working_directory}")
+      end
+
+      def delete_files(files_to_delete: [], custom_message: nil)
+        files_to_delete.each do |current_file|
+          target_path = current_file.gsub(self.working_directory + "/", "")
+          file = bucket.file(target_path)
+          UI.message("Deleting '#{target_path}' from Google Cloud Storage bucket '#{self.bucket_name}'...")
+          file.delete
+        end
+        finished_pushing_message
+      end
+
+      def upload_files(files_to_upload: [], custom_message: nil)
+        # `files_to_upload` is an array of files that need to be uploaded to Google Cloud
+        # Those doesn't mean they're new, it might just be they're changed
+        # Either way, we'll upload them using the same technique
+
+        files_to_upload.map do |current_file|
```

Comment:
this map is not returned by `upload_files` Just raising it up. If that's intentionally than maybe `each` would be more appropriate.

Reply:
will update

Reply:
Yep was a copy & paste error